Zend/zend_compile.c

```diff
@@ -6737,12 +6737,38 @@ zend_op *zend_compile_class_decl(zend_ast *ast, zend_bool toplevel) /* {{{ */
 			zend_error_noreturn(E_COMPILE_ERROR, "Clone method %s::%s() cannot be static",
 				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
 		} else if (ce->clone->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+			if (ZEND_TYPE_PURE_MASK(ce->clone->common.arg_info[-1].type) != MAY_BE_VOID) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Method %s::%s() must return void",
+					ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+			}
+		}
+	}
+
+	if (ce->__set && ce->__set->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_PURE_MASK(ce->__set->common.arg_info[-1].type) != MAY_BE_VOID) {
 			zend_error_noreturn(E_COMPILE_ERROR,
-				"Clone method %s::%s() cannot declare a return type",
-				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+				"Setter %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__set->common.function_name));
 		}
 	}
 
+	if (ce->__unset && ce->__unset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_PURE_MASK(ce->__unset->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Method %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__unset->common.function_name));
+		}
+	}
+
+	if (ce->__isset && ce->__isset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		 if (ZEND_TYPE_PURE_MASK(ce->__isset->common.arg_info[-1].type) != MAY_BE_BOOL) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				 "Method %s::%s() must return a boolean value",
+				 ZSTR_VAL(ce->name), ZSTR_VAL(ce->__isset->common.function_name));
+		 }
+	}
+
```

Comment:
@nikic Accessing `ce->serialize_func` here has no effect. Is there a way to check the `__serialize` method here?

Reply:
I think you have to get the function from the function table, see https://heap.space/xref/php-src/ext/standard/var.c?r=73d02c3b#994.

Reply:
@cmb69 It worked, thanks!